wandb/__init__.py

```diff
@@ -508,7 +516,38 @@ def __exit__(self, *args):
     return Monitor(options)
 
 
-def log(row=None, commit=True, step=None, *args, **kwargs):
+_async_log_queue = queue.Queue()
+_async_log_thread_shutdown_event = threading.Event()
+_async_log_thread_complete_event = threading.Event()
+_async_log_thread = None
+def _async_log_thread_target():
+    """Consumes async logs from our _async_log_queue and actually logs them"""
+    global _async_log_thread
+    shutdown_requested = False
+    while not shutdown_requested:
+        try:
+            kwargs = _async_log_queue.get(True)
+            log(**kwargs)
+        except queue.Empty:
+            shutdown_requested = _async_log_thread_shutdown_event.wait(1) and _async_log_queue.empty()
+    _async_log_thread_complete_event.set()
+    _async_log_thread = None
+
+def _ensure_async_log_thread_started():
+    """Ensures our log consuming thread is started"""
+    global _async_log_thread
+
+    if _async_log_thread is None:
+        _async_log_thread = threading.Thread(target=_async_log_thread_target)
+        _async_log_thread.daemon = True
+        _async_log_thread.start()
+
+def shutdown_async_log_thread():
+    """Shuts down our async logging thread"""
+    _async_log_thread_shutdown_event.set()
+    _async_log_thread_complete_event.wait(1)
```

Comment:
this wait should be something larger than 1, since that is the delay for empty queue detection.

Also you could check the return value and print a warning if the wait times out (returns None)